Refactor/remove mrouter

morantr opened the pull request:
Remove Controllers Mrouter from the code.

tomereli commented on controller/src/beerocks/master/beerocks_master_main.cpp:
```diff
@@ -304,7 +304,14 @@ int main(int argc, char *argv[])
     std::string master_uds  = beerocks_master_conf.temp_path + std::string(BEEROCKS_MASTER_UDS);
     std::string mrouter_uds = beerocks_master_conf.temp_path + std::string(BEEROCKS_MROUTER_UDS);
 
-    son::db master_db(master_conf, logger);
```
Why not just use `beerocks::net::network_utils::get_iface_info(bridge_info, bridge_iface)` when we need it?

morantr replied:
Because we only need to read it once. There is no point in reading it all over again every time.

morantr commented:
For some reason the test_flows.sh fails on some steering tests. I tried to run it on local GW and it worked for me. So I couldn't find the reason why it is not working.

arnout commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -748,6 +758,8 @@ class db {
     std::shared_ptr<uint8_t> certification_tx_buffer;
     std::unordered_map<sMacAddr, std::list<bss_info_conf_t>> bss_infos; // key=al_mac
+
+    master_thread *m_master_thread_ctx;
```
Instead of making this a pointer, I'd make it a reference and pass it in the constructor (obviously) instead of set_master_thread_ctx(). This has the following advantages:

- Can remove the set function;
- It's guaranteed not to change over the lifetime;
- It indicates that the lifetime of db is tied to the lifetime of masther_thread;
- get_master_thread_ctx() returns a reference, so no need to check nullptr.

morantr replied:
It can't be done since the database is being created before the son_master is created and being passed as a reference to it.
I guess there is a solution for that, like not creating the database outside to the son_master thread and initialize it in the son_master constructor.
Do you want me to do it? It will need an extra commit on top.
@arnout @tomereli

tomereli replied:
Sounds legit - the son master is the "owner" of the DB so it makes sense that it will create it in its constructor, so I would say go for it.

arnout replied:
I'm OK either way. It was anyway just a nitpick comment.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -112,7 +138,7 @@ bool master_thread::init()
     }
 
-    return socket_thread::init();
+    return true;
```
This change should have been part of "controller: Make son_master_thread to be derived from btl" commit.

morantr replied:
what? the returning of `true`? it is not critical...

morantr replied:
Not critical.

arnout commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -136,7 +136,8 @@ class db {
 
-    db(sDbMasterConfig &config_, beerocks::logging &logger_) : config(config_), logger(logger_)
+    db(sDbMasterConfig &config_, beerocks::logging &logger_, const std::string &local_bridge_mac)
+        : config(config_), logger(logger_), m_local_bridge_mac(local_bridge_mac)
```
Although this address is indeed the local bridge address at the moment, we will have to change this at some point (because it's wrong: it should be a locally administered address according to 1905.1, while the bridge address is usually a OUI address). Therefore, I wouldn't create a new variable that has "bridge" in the name. Instead, use `al_mac`, `m_al_mac` etc.

tomereli replied:
+1 for `m_al_mac`

morantr replied:
I don't understand, in the future `al_mac` and the bridge mac won't be equal?

tomereli replied:
The name should be al_mac. Each device can define its own al mac, currently it is the bridge mac but it can change, so the answer is "maybe".

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -187,6 +187,42 @@ bool master_thread::socket_disconnected(Socket *sd)
 
 bool master_thread::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx)
 {
+    // Note: sd can be equal to nullptr it the master_thread is derived from btl_local_bus and the
```
The part `it the master_thread is derived from btl_local_bus` is not clear and IMO redundant.  Actually, the entire comment is redundant - we're only using `sd` if `from_bus` returns false, so all is well...

morantr replied:
Removed the comment.

arnout commented on controller/src/beerocks/master/son_actions.cpp:
```diff
@@ -294,23 +296,18 @@ void son_actions::handle_dead_node(std::string mac, std::string hostap_mac, db &
     if ((mac_type == beerocks::TYPE_IRE_BACKHAUL || mac_type == beerocks::TYPE_CLIENT) &&
         database.is_node_wireless(mac)) {
-        Socket *hostap_mac_sd = database.get_node_socket(hostap_mac);
-        if (hostap_mac_sd) {
```
We still need this condition, but then on the MAC address:

```
if (!agent_mac.empty()) {
```

Or actually, looking at the rest, we don't do this anywhere else, so probably not needed. In that case, it should be mentoined in the commit message. Which is anyway is not entirely clear, better would be:

```
 controller: send_cmdu_to_agent: use bus instead of mrouter socket

Change "send_cmdu_to_agent" on son_actions to send the message to the
Bus instead of UDS from mrouter.

To replace the Socket, we need the master_thread as an argument to be able
to call the send_cmdu_to_bus function, and the MAC address of the agent.
Convert all callers to get the MAC address of the agent from the database 
with get_node_parent_ire instead of getting the Socket with get_node_socket.

There were four cases where the return value of get_node_socket() was
checked. Remove that check now, because we don't do it anywhere else: a
hostap node is guaranteed to have a ire parent, and if the MAC address is
empty, it will fail in the sending stage.
```

morantr replied:
Will change the commit message

arnout commented on controller/src/beerocks/master/tasks/client_steering_task.cpp:
```diff
@@ -87,7 +87,6 @@ void client_steering_task::work()
 
 void client_steering_task::steer_sta()
 {
-    Socket *sta_socket = database.get_node_socket(sta_mac);
```
This one is different from the rest of the changes, so it should be mentioned explicitly in the commit message. Actually, I even think it could be a separate commit.

morantr replied:
It really not. This is used for backhaul roaming. Here we get the backhaul socket which is basically the same mac address of the parent bridge mac which we later get.
On line 144 I replace the validation of this socket by comparing the node type to `beerocks::TYPE_IRE_BACKHAUL`.
Since I remove all `get_node_socket` calls in this commit, this needs to be part of this commit/

arnout reviewed (changes requested):
Reviewed up to "controller: Change "send_cmdu_to_agent" on son_actions "

Nothing major so far, it's definitely going in the right direction.

Are there any cleanups that can still be done but haven't been taken in this PR? If yes, it would be nice to mention this as a comment in the PR itself.

arnout commented on agent/src/beerocks/slave/agent_ucc_listener.cpp:
```diff
@@ -64,16 +64,16 @@ void agent_ucc_listener::clear_configuration()
     // On the agent side, the dest_alid is not really needed since the destination socket will
     // always be the controller socket.
-    return *m_controller_sd;
+    return (*m_controller_sd)->getPeerMac() == dest_alid;
```
Nice! I hadn't thought of the possibility to validate the destination.

arnout reviewed (approved):
Reviewed the rest of the PR.

Since my comments in the first review were all of the nitpicky type, I'm approving already.

arnout commented:
One more thing: `transport_socket_thread::init()` must be done already in commit `controller: Make son_master_thread to be derived from btl` otherwise the UCC class will trigger an error because it tries to add a socket when the poller isn't initialized yet.

morantr committed:
btl: Add Bus socket server to btl_uds

Preparative commit.
When using UDS BUS (MSGLIB=None), the Mrouter can have only one
incoming UDS connection which is from the local Agent. On that
connection event, the Mrouter calls "bus_connected" function which
marks the incoming socket as BUS socket.
When the son_master will be derived from btl_uds, more than one socket
could get connected, since also the BML/CLI are open UDS as well, so
we could not use "bus_connected" function upon "socket_connected"
event.
To solve this, a second server socket was added to btl_uds, with a
different UDS address. That way the bus socket will be marked only on
the socket that was opened on the new socket server.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
btl: Move "send_cmdu_to_bus()" function to public

Preparative commit.
Move the function "send_cmdu_to_bus()" to public to allow external
classes (such as son_action) to send cmdu on the bus.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Make son_master_thread to be derived from btl

Make son_master_thread to be derived from transport_socket_thread.
Also, change the constructor to get "master_uds" as a const reference.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Saving son_master context into database

Preparative commit.
Add setter and getter of son_master context into the database and set
it in the son_master init function.
This is needed since the  send_cmdu_to_bus function is within the
btl, and in order to allow external classes to send a CMDU to the bus,
saving the son_master context is needed.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Initialize the bus from son_master init function

Add initialization of  the bus from son_master init function and
subscribe to the messages the controller needs to receive.
This change is copied "as is" from the Mrouter.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Add local bridge mac to the database

Preparative commit.
Add a setter to local bridge mac on database constructor and a getter
function to the database as well.
This is required because sending CMDU to the bus requires to set
the source al_mac address.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Remove the removal of disconnected agents from the database

Remove the removal of agent and sub-agent nodes from socket_disconnected
function on the son_master. Thus since the agents cannot disconnect
anymore because the controller is linked through the Bus and doesn't
hold UDS anymore.

Removed the related database helper function as well.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Add message filters to "handle_cmdu" function on son_master

Copy Bus message filters from Mrouter to the son_master "handle_cmdu"
function.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: send_cmdu_to_agent: use bus instead of mrouter socket

Change "send_cmdu_to_agent" on son_actions to send the message to the
Bus instead of UDS from mrouter.

To replace the Socket, we need the master_thread as an argument to be able
to call the send_cmdu_to_bus function, and the MAC address of the agent.
Convert all callers to get the MAC address of the agent from the database
with get_node_parent_ire instead of getting the Socket with get_node_socket.

There were four cases where the return value of get_node_socket() was
checked. Remove that check now, because we don't do it anywhere else: a
hostap node is guaranteed to have a ire parent, and if the MAC address is
empty, it will fail in the sending stage.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
ucc_listener: Add function "validate_destination_alid"

Add to ucc_listener the function "validate_destination_alid" which
validated the "alid_mac" received from the ucc "den_send_1905" command.
Replace this function with "get_dev_send_1905_destination_socket"
function which is not relevant anymore no that the controller uses the
Bus directly.
Also, change "send_cmdu_to_destination" function arguments to get
distination alid MAC address instead of socket.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Remove the Mrouter

Remove the Mrouter from the code and all related code and references
to it as well.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Remove node socket from the database

Remove node socket from the database including its setter and getter,
and all of its calling.
Saving the node socket is not relevant anymore now that the controller
is connected directly to the Bus.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Add missing memory allocations error checks

Add missing memory allocations error checks and changed the error level
to FATAL where there was an error check.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
btl: Add Bus to the Poller only once

Previously the Bus was added to the Poller on each topic that we were
subscribing to the Bus. Doing it is redundant and causing multiple
irrelevant prints at the beginning of the log.
Moved the adding of the Bus to the Poller, after we have subscribed
to all the topics and removed the call to Sync function since it is
only relevant to call it if we changed the bus subscription after we
added it to the Poller which is not the case anymore.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
bcl: Fix Controller crash

After converting the son_master to BTL, it was observed that the
test_flow.sh did not pass anymore and the CI always failed.
This problem existed all the time, and not been covered since until
now, the BTL socket thread didn't have any UDS connecting/disconnecting
like the son_master has - BML socket.
When socket disconnect and handled in the socket_thread, the calling to
"CloseSocket" has been putting the file descriptor of the socket
to '-1', and when then on "remove_socket" the poller has received
fd='-1', which caused unexpected behavior.

Fix this by replacing the order of the calling to "remove_socket" and
"CloseSocket".

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

tomereli commented on common/beerocks/btl/btl_local_bus.cpp:
```diff
@@ -82,16 +71,26 @@ bool transport_socket_thread::bus_init()
-
     LOG_IF(!bus, FATAL) << "Bus is not allocated!";
     LOG_IF(!poller, FATAL) << "Poller is not allocaed!";
     bus->Init();
```
I think we should have a `bus_Init()` which will call `bus->init()` so in the future if we call `bus_subscribe` more than once it will not cause an issue.

morantr replied:
we have it already.
